xdsl/frontend/passes/promoters.py

```diff
@@ -0,0 +1,315 @@
+from dataclasses import dataclass, field
+from typing import List
+from xdsl.frontend import symref
+from xdsl.frontend.exception import FrontendProgramException
+from xdsl.ir import Block, Operation, Region, SSAValue
+from xdsl.rewriter import Rewriter
+from xdsl.dialects import affine, scf
+
+# Promoters
+# =========
+#
+# Promoters are extremely important for desymrification of an operation - they
+# dictate how an operation updates values. Generally speaking, for any operation
+# in the form that uses some symbols `@a`, `@b` and `@c`:
+#
+# ```
+# dialect.op {
+#   use @a, @b, @c
+# }
+# ```
+#
+# promoter for 'dialect.op' rewrites it into:
+#
+# ```
+# fetch @a, @b, @c
+# %new_symbol_values = dialect.op [%argument_symbol_values] {
+#   // no symbols used
+# }
+# update @a, @b, @c with %new_symbol_values
+# ```
+#
+# This way the regions inside an operation do not access any symbols, and any
+# symbol updates are propogated outside of the operation.
+
+
+def insert_after(op: Operation, new_op: Operation):
+    """Inserts a new operation after another operation."""
+    op_parent_block = op.parent_block()
+    op_idx = op_parent_block.get_operation_index(op)
+    op_parent_block.insert_op(new_op, op_idx + 1)
+
+
+def insert_before(op: Operation, new_op: Operation):
+    """Inserts a new operation before another operation."""
+    op_parent_block = op.parent_block()
+    op_idx = op_parent_block.get_operation_index(op)
+    op_parent_block.insert_op(new_op, op_idx)
+
+
+def is_read_only(symbol: str, block: Block) -> bool:
+    """
+    Returns true if the symbol is never updated.
+    """
+    return all([isinstance(op, symref.Update) and op.symbol.data.data == symbol for op in block.ops])
+
+
+def get_fetch_op(symbol: str, block: Block) -> None | symref.Fetch:
+    """
+    Returns a fetch of this symbol in this block. Should be used when the symbol
+    is used fetched in the block. Returns None if the symbol is not found.
+    """
+    for op in block.ops:
+        if isinstance(
+                op,
+                symref.Fetch) and op.symbol.data.data == symbol:
+            return op
+    return None
+
+
+def get_update_op(symbol: str, block: Block) -> None | symref.Update:
+    """
+    Returns an update to this symbol in this block. Should be used when the
+    symbol is updated once in the block. Returns None if the symbol is not
+    found.
+    """
+    for op in block.ops:
+        if isinstance(
+                op,
+                symref.Update) and op.symbol.data.data == symbol:
+            return op
+    return None
+
+
+@dataclass
+class Promoter:
+    """Base promoter class."""
+
+    op: Operation
+    """Operation for promotion."""
+
+    rewriter: Rewriter = field(default_factory=Rewriter)
+    """Rewriter used to erase or replace ops."""
+
+    def _find_symbols(self) -> List[str]:
+        """Returns a set of all unpromoted symbols."""
+        symbols = []
```

Comment:
This also could be a bit neater with a list comprehension
```suggestion
        symbols = list(set(
            op.symbol.data.data
            if isinstance(op, symref.Fetch | symref.Update)
            for op in block.ops
            for block in region.blocks
            for region in self.op.regions
        ))
```

Reply:
I don't mind having this but this nested structure seems very ugly to me :)